pybamm/models/base_model.py

```diff
@@ -188,17 +188,26 @@ def update(self, *submodels):
             The submodels from which to create new model
         """
         for submodel in submodels:
-            # check for duplicates in keys
-            vars = [var.id for var in submodel.rhs.keys()] + [
-                var.id for var in self.rhs.keys()
-            ]
-            assert len(vars) == len(set(vars)), pybamm.ModelError("duplicate variables")
 
-            # update dicts
-            self._rhs.update(submodel.rhs)
-            self._initial_conditions.update(submodel.initial_conditions)
-            self._boundary_conditions.update(submodel.boundary_conditions)
-            self._variables.update(submodel.variables)
+            # check and then update dicts
+            self.check_and_combine_dict(self._rhs, submodel.rhs)
+            self.check_and_combine_dict(self._algebraic, submodel.algebraic)
+            self.check_and_combine_dict(
+                self._initial_conditions, submodel.initial_conditions
+            )
+            self.check_and_combine_dict(
+                self._boundary_conditions, submodel.boundary_conditions
+            )
+            self._variables.update(submodel.variables)  # keys are strings so no check
```

Comment:
Could check that strings aren't being repeated (and hence overwritten) but not crucial